src/Features/Core/Completion/AbstractCompletionService.CompletionRules.cs

```diff
@@ -8,17 +9,33 @@ namespace Microsoft.CodeAnalysis.Completion
 {
     internal partial class AbstractCompletionService
     {
-        private class CompletionRules : ICompletionRules
+        protected class CompletionRules : ICompletionRules
         {
+            private readonly object _gate = new object();
             private readonly AbstractCompletionService _completionService;
-            private readonly PatternMatcher _patternMatcher = new PatternMatcher(verbatimIdentifierPrefixIsWordCharacter: true);
+            private readonly Dictionary<string, PatternMatcher> patternMatcherMap = new Dictionary<string, PatternMatcher>();
```

Comment:
Completion still works in an interesting manner.   Because the outer loop is in the completion service, and the inner check is in the 'compleiton rules', there is no way to share a pattern matcher normally across calls to the MatchesFilterText.

one way to address this would be to change MatchesFilterText to take a PatternMatcher.  But that's odd itself.  Now the completion service needs to know that MatchesFilterText uses a PatternMatcher.  

That all seemed unclean, so instead we just cache the pattern matcher we use for any filter text.  This effectively means that completion has the same caching of pattern matcher data that we had before (it's just moved out from teh pattern matcher to the completion infrastructure).   This isn't great, but at least NavigateTo and Progression are both much nicer and do much less work.